controller/watch/watcher.go

```diff
@@ -249,48 +249,65 @@ func (w *Watcher) Watch(t resourceType, ns string, wcb watchCallback) error {
 }
 
 func deploymentWatcher(di deployments.DeploymentInterface, wcb watchCallback) {
-	for {
-		log.Println("List all existing Deployments")
-		// First do List on the resource to bring things up to date.
-		l, err := di.List(api.ListOptions{})
-		for _, d := range l.Items {
-			log.Printf("Found Deployment name: %s\n", d.Name)
-			event := watch.Event{
-				Type:   watch.Added,
-				Object: &d,
+	// 'listChan' is our signal to know if we should re-get the full list
+	// of resources.
+	listChan := make(chan bool, 1)
+
+	go func() {
```

Comment:
>the go routine isn't trying to make any guarantees

Why not? I thought the goal was to list before the watch, while not missing any watch events while the list was occurring. Doing that requires a happens-before guarantee.

>I need a go routine because the two things (list() vs watcher()) are run in parallel

It seems that a simpler way to solve the problem of avoiding the race is to open (1) the watch channel, (2) list, and (3) start consuming on the watch channel?